tmk_core/common/action_util.c

```diff
@@ -364,3 +365,120 @@ __attribute__((weak)) void oneshot_layer_changed_kb(uint8_t layer) { oneshot_lay
  * FIXME: needs doc
  */
 uint8_t has_anymod(void) { return bitpop(real_mods); }
+
+/** \brief custom modifier handling
+ *
+ * This function checks if the given modifier mask is set and the given layer is
+ * the current spec. If not it returns true. If the mask is set and the key
+ * pressed, it deregisters the currently held modifiers, executes the given
+ * function and returns the value returned by the function. If the key is
+ * released, it simply runs the function and returns its value.
+ *
+ * This function is meant to be used in process_record_user to be able to
+ * trigger specific action on key code with modifier, for example when shift-esc
+ * is pressed.
+ *
+ * \param modifier_mask modifier mask to check against current modifiers held.
+ * \param layer layer to check against current layer.
+ * \param keycode the processed keycode, will be forwarded to handler.
+ * \param record the processed record, will be forwarded to handler.
+ * \param data anonymous data to be passed to the handler.
+ * \param handler to be called when modifiers and layer are matched.
+ * \return true if modifiers or layer did not match, return value of handler otherwise.
+ */
+bool run_on_modifiers(uint8_t modifier_mask, int16_t layer, uint16_t keycode,  keyrecord_t *record,
+                      void *data,  bool (*handler)(uint16_t keycode,  keyrecord_t *record, void *data)) {
```

Comment:
`int16_t layer` may be made `int8_t` (layer numbers are limited to the 0…31 range, because `uint32_t` is used as a bit mask type, and most of the core code which deals with layer numbers uses `uint8_t`). This can save some resources on AVR.

Although I'm not sure whether such layer state testing should be encouraged at all — it would probably be better to put the appropriate keycodes into the keymap instead, so that the handler with `run_on_modifiers()` or `custom_keycode_on_modifiers()` would be invoked only when pressing the key while on the appropriate layer. It would also remove any other considerations, such as whether a higher layer is active and should override the key behavior, or whether the layer 0 is always active or not, or whether `default_layer_state` should be checked too (`layer_state_is()` does not look at it).